arangod/RocksDBEngine/Methods/RocksDBTrxBaseMethods.cpp

```diff
@@ -0,0 +1,408 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2021 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Manuel Pöter
+////////////////////////////////////////////////////////////////////////////////
+
+#include "RocksDBTrxBaseMethods.h"
+
+#include "Random/RandomGenerator.h"
+#include "RocksDBEngine/RocksDBLogValue.h"
+#include "RocksDBEngine/RocksDBSettingsManager.h"
+#include "RocksDBEngine/RocksDBSyncThread.h"
+#include "RocksDBEngine/RocksDBTransactionState.h"
+#include "Statistics/ServerStatistics.h"
+
+#include <rocksdb/utilities/write_batch_with_index.h>
+
+using namespace arangodb;
+
+RocksDBTrxBaseMethods::RocksDBTrxBaseMethods(RocksDBTransactionState* state, rocksdb::TransactionDB* db)
+    : RocksDBTransactionMethods(state),
+      _db(db) {
+  TRI_ASSERT(!_state->isReadOnlyTransaction());
+  _readOptions.prefix_same_as_start = true;  // should always be true
+  _readOptions.fill_cache = _state->options().fillBlockCache;
+}
+
+RocksDBTrxBaseMethods::~RocksDBTrxBaseMethods() {
+  cleanupTransaction();
+}
+
+bool RocksDBTrxBaseMethods::DisableIndexing() {
+  if (!_indexingDisabled) {
+    TRI_ASSERT(_rocksTransaction);
+    _rocksTransaction->DisableIndexing();
+    _indexingDisabled = true;
+    return true;
+  }
+  return false;
+}
+
+bool RocksDBTrxBaseMethods::EnableIndexing() {
+  if (_indexingDisabled) {
+    TRI_ASSERT(_rocksTransaction);
+    _rocksTransaction->EnableIndexing();
+    _indexingDisabled = false;
+    return true;
+  }
+  return false;
+}
+
+Result RocksDBTrxBaseMethods::beginTransaction() {
+  TRI_ASSERT(_rocksTransaction == nullptr);
+
+  createTransaction();
+  TRI_ASSERT(_rocksTransaction != nullptr);
+  _readOptions.snapshot = _rocksTransaction->GetSnapshot();
+
+  return {};
+}
+
+Result RocksDBTrxBaseMethods::commitTransaction() {
+  Result result = doCommit();
+  if (result.ok()) {
+    cleanupTransaction();
+  }
+  return result;
+}
+
+Result RocksDBTrxBaseMethods::abortTransaction() {
+  Result result;
+  if (_rocksTransaction != nullptr) {
+    rocksdb::Status status = _rocksTransaction->Rollback();
+    result = rocksutils::convertStatus(status);
+  }
```

Comment:
This looks kind of asymmetrical:
When we commit a transaction, we call `cleanupTransaction()` if committing succeeds.
If committing fails or when we abort a transaction, we don't call `cleanupTransaction()` at all. It is then only called in the class' destructor. 
This may be ok. I was just wondering.